GPT/beta-commands/gpt-function-calling.py

```diff
@@ -0,0 +1,161 @@
+                    "parameters": {
+                        "type": "object",
+                        "properties": {
+                            "str": {
+                                "type": "string",
+                                "description": "The text to display",
+                            }
+                        },
+                        "required": ["str"],
+                    },
+                },
+            },
+            {
+                "type": "function",
+                "function": {
+                    "name": "search_for_command",
+                    "description": "search_for_command(str: string) - this searches for a command in the VSCode command palette. If I ask you to do something, please use this command to search for an appropriate command.",
+                    "parameters": {
+                        "type": "object",
+                        "properties": {
+                            "str": {
+                                "type": "string",
+                                "description": "The command to search for",
+                            }
+                        },
+                        "required": ["str"],
+                    },
+                },
+            }
+        ],
+        "max_tokens": 2024,
+        "temperature": 0.6,
+        "n": 1,
+        "stop": None,
+        "model": settings.get("user.openai_model"),
+    }
+
+
+    response = requests.post(url, headers=headers, data=json.dumps(data))
+
+    message = response.json()["choices"][0]["message"]
+    if response.status_code == 200:
+        notify("GPT Task Completed")
+        print(response.json())
+        process_function_calls(insert_response, message)
+
+        content = (message["content"] or "").strip()
+        if len(content) != 0:
+            display_response(content)
+
+    else:
+        notify("GPT Failure: Check API Key, Model, or Prompt")
+        print(response.json())
+
+def process_function_calls(insert_response, message):
+    try:
+        tool_calls = message["tool_calls"]
+        while tool_calls:
+            tool = tool_calls.pop()
+            first_argument = tool['function']['arguments']
+            try:
+                first_argument = json.loads(first_argument)['str']
+            except Exception as e:
+                notify(f"Argument Jason was malformed: {e}")
+            match tool['function']['name']:
+                case 'insert':
+                    insert_response(first_argument)
+                case 'display':
+                    display_response(first_argument)
+                case 'notify':
+                    notify_user(first_argument)
+                case 'search_for_command':
+                    search_for_command(first_argument)
+    except Exception as e:
+        notify(f"No tool_calls found in response from LLM: {e}")
```

Comment:
we prob want this exception to wrap just the statement where we extract tool_calls and `return` from here